src/main/java/io/vertx/core/impl/FutureImpl.java

```diff
@@ -78,29 +80,48 @@ public synchronized boolean isComplete() {
     return failed || succeeded;
   }
 
-  /**
-   * Set a handler for the result. It will get called when it's complete
-   */
-  public Future<T> setHandler(Handler<AsyncResult<T>> handler) {
-    boolean callHandler;
+  @Override
+  public Future<T> onComplete(Handler<AsyncResult<T>> h) {
+    if (h == null) {
+      throw new NullPointerException("No null handler accepted");
+    }
     synchronized (this) {
-      if (handler == null) {
-        this.handler = null;
-        return this;
-      } else {
-        callHandler = isComplete();
-        if (!callHandler) {
-          this.handler = handler;
+      if (!isComplete()) {
+        if (handler == null) {
+          handler = h;
+        } else {
+          addHandler(h);
         }
+        return this;
       }
     }
-    if (callHandler) {
-      dispatch(handler);
-    }
+    dispatch(h);
     return this;
   }
 
+  private void addHandler(Handler<AsyncResult<T>> h) {
+    Handlers<T> handlers;
+    if (handler instanceof Handlers) {
+      handlers = (Handlers<T>) handler;
+    } else {
+      handlers = new Handlers<>();
+      handlers.add(handler);
+      handler = handlers;
+    }
+    handlers.add(h);
+  }
+
   protected void dispatch(Handler<AsyncResult<T>> handler) {
+    if (handler instanceof Handlers) {
+      for (Handler<AsyncResult<T>> h : (Handlers<T>)handler) {
+        dispatch0(h);
+      }
+    } else {
+      dispatch0(handler);
+    }
+  }
+
+  private void dispatch0(Handler<AsyncResult<T>> handler) {
```

Comment:
Any more meaningful name?

Reply:
no :-)

Reply:
the goal is to avoid recursive dispatch that would not be jitted

Reply:
I can rename it to `doDispatch` ?

Reply:
Yep